File: collection/2020 hw2 java polymorphism/shapes/Rectangle.java
import java.util.ArrayList;
import java.util.List;

public class Rectangle extends Quadrilateral implements SymmetricTwoDShape {

    //private TwoDPoint[] vertices = new TwoDPoint[4];

    public Rectangle(List<TwoDPoint> vertices) {
        super(vertices);
    }
    
    public TwoDPoint[] getVertices() {
        return super.getVertices();
    }

	/**
     * The center of a rectangle is calculated to be the point of intersection of
     * its diagonals.
     *
     * @return the center of this rectangle.
     */
    @Override
    public Point center() {
        TwoDPoint point1 = this.getVertices()[0];
        TwoDPoint point2 = this.getVertices()[2];
        
        TwoDPoint centerPoint = new TwoDPoint(
            (point1.coordinates()[0] + point2.coordinates()[0] / 2),
            (point1.coordinates()[1] + point2.coordinates()[1] / 2));
        return centerPoint;
    }

    @Override
    public boolean isMember(List<? extends Point> vertices) {
        TwoDPoint point1 = (TwoDPoint)vertices.get(0);
        TwoDPoint point2 = (TwoDPoint)vertices.get(1);
        TwoDPoint point3 = (TwoDPoint)vertices.get(2);
        TwoDPoint point4 = (TwoDPoint)vertices.get(3);
        /* List<TwoDPoint> toSet = new ArrayList<TwoDPoint>();
        toSet.add(point1);
        toSet.add(point2);
        toSet.add(point3);
        toSet.add(point4);
        Rectangle toCheck = new Rectangle(toSet);
        double[] sideLengths = toCheck.getSideLengths(); 
        if (sideLengths[0] == 0 || sideLengths[1] == 0 || sideLengths[2] == 0 || sideLengths[3] == 0) {
            return false;
        }*/
        
        double oneDiag = Math.sqrt(Math.pow((point1.coordinates()[0] - point3.coordinates()[0]), 2) +
                         Math.pow((point1.coordinates()[1] - point3.coordinates()[1]), 2));
        double twoDiag = Math.sqrt(Math.pow(point2.coordinates()[0] - point4.coordinates()[0], 2) +
                         Math.pow((point2.coordinates()[1] - point4.coordinates()[1]), 2));
        return (oneDiag == twoDiag);
    }

    private boolean isTheSamePoint(TwoDPoint arg1, TwoDPoint arg2) {
        if (arg1.getXcor() == arg2.getXcor() && arg1.getYcor() == arg2.getYcor()) {
            return true;
        }
        else {
            return false;
        }
    }

    @Override
    public double area() {
        double toReturn;
        double side1 = Math.sqrt(Math.pow((getVertices()[0].coordinates()[0] - getVertices()[1].coordinates()[0]), 2) +
                                  Math.pow((getVertices()[0].coordinates()[1] - getVertices()[1].coordinates()[1]), 2));
        double side2 = Math.sqrt(Math.pow((getVertices()[1].coordinates()[0] - getVertices()[2].coordinates()[0]), 2) +
                                   Math.pow((getVertices()[1].coordinates()[1] - getVertices()[2].coordinates()[1]), 2));
        toReturn = side1 * side2;
        return toReturn;
    }
}

File: collection/2020 hw3 python/rectangle.py
import unittest
from quadrilateral import Quadrilateral
from two_d_point import TwoDPoint
import math

class Rectangle(Quadrilateral):

    def __init__(self, *floats):
        super().__init__(*floats)
        if not self.__is_member():
            raise TypeError("A rectangle cannot be formed by the given coordinates.")

    def __is_member(self):
        """Returns True if the given coordinates form a valid rectangle, and False otherwise."""
        point0 = self.vertices[0]
        point1 = self.vertices[1]
        point2 = self.vertices[2]
        point3 = self.vertices[3]
        
        oneDiag = math.sqrt(math.pow(point0.x - point2.x, 2) + math.pow(point0.y - point2.y, 2))
        twoDiag = math.sqrt(math.pow(point1.x - point3.x, 2) + math.pow(point1.y - point3.y, 2))
        return (oneDiag == twoDiag)

    def center(self):
        """Returns the center of this rectangle, calculated to be the point of intersection of its diagonals."""
        point0 = self.vertices[0]
        point1 = self.vertices[1]
        point2 = self.vertices[2]
        point3 = self.vertices[3]
        midpointX = (point0.x + point2.x)/2
        midpointY = (point0.y + point2.y)/2
        pointToReturn = TwoDPoint(midpointX, midpointY)
        return pointToReturn

    def area(self):
        """Returns the area of this rectangle. The implementation invokes the side_lengths() method from the superclass,
        and computes the product of this rectangle's length and width."""
        sidelengths = self.side_lengths()
        areaToReturn = sidelengths[0] * sidelengths[1]
        return areaToReturn

    def __eq__(self, other: object) -> bool:
        """
        Two Rectangles are equal if their vertices are equal
        """
        if type(other) is Rectangle:
            vertex0 = self.vertices[0]
            vertex1 = self.vertices[1]
            vertex2 = self.vertices[2]
            vertex3 = self.vertices[3]

            compare_vertex0 = other.vertices[0]
            compare_vertex1 = other.vertices[1]
            compare_vertex2 = other.vertices[2]
            compare_vertex3 = other.vertices[3]
        
            return (vertex0 == compare_vertex0 and vertex1 == compare_vertex1 and vertex2 == compare_vertex2 and vertex3 == compare_vertex3)
        else:
            return False

    def __str__(self) -> str:
        return 'Rectangle\n(%g, %g)\n(%g, %g)\n(%g, %g)\n(%g, %g)' % (self.vertices()[0].x, self.vertices()[0].y, self.vertices()[1].x, self.vertices()[1].y, self.vertices()[2].x, self.vertices()[2].y, self.vertices()[3].x, self.vertices()[3].y)